cpp/src/io/json/json_tree.cu

```diff
@@ -0,0 +1,670 @@
+    b. Convert the node categories to node types.
+       Node type is defined as node category enum value if it is not a field node,
+       otherwise it is the unique node id assigned by the hashmap (value shifted by #NUM_CATEGORY).
+  2. Preprocessing: Translate parent node ids after sorting by level.
+    a. sort by level
+    b. get gather map of sorted indices
+    c. translate parent_node_ids to new sorted indices
+  3. Find level boundaries.
+     copy_if index of first unique values of sorted levels.
+  4. Per-Level Processing: Propagate parent node ids for each level.
+    For each level,
+      a. gather col_id from previous level results. input=col_id, gather_map is parent_indices.
+      b. stable sort by {parent_col_id, node_type}
+      c. scan sum of unique {parent_col_id, node_type}
+      d. scatter the col_id back to stable node_level order (using scatter_indices)
+    Restore original node_id order
+  5. Generate row_offset.
+    a. stable_sort by parent_col_id.
+    b. scan_by_key {parent_col_id} (required only on nodes who's parent is list)
+    c. propagate to non-list leaves from parent list node by recursion
+  **/
+  // GPU version
+  // 1. Convert node_category+fieldname to node_type.
+  using hash_table_allocator_type = rmm::mr::stream_allocator_adaptor<default_allocator<char>>;
+  using hash_map_type =
+    cuco::static_map<size_type, size_type, cuda::thread_scope_device, hash_table_allocator_type>;
+
+  constexpr size_type empty_node_index_sentinel = std::numeric_limits<size_type>::max();
+  auto num_nodes                                = d_tree.node_categories.size();
+  hash_map_type key_map{compute_hash_table_size(num_nodes),  // TODO reduce oversubscription
+                        cuco::sentinel::empty_key{empty_node_index_sentinel},
+                        cuco::sentinel::empty_value{empty_node_index_sentinel},
+                        hash_table_allocator_type{default_allocator<char>{}, stream},
+                        stream.value()};
+  auto d_hasher = [d_input          = d_input.data(),
+                   node_range_begin = d_tree.node_range_begin.data(),
+                   node_range_end   = d_tree.node_range_end.data()] __device__(auto node_id) {
+    // TODO if node_category is hashed, then no need for transform later. for field only, use string
+    // hash too.
+    auto field_name = cudf::string_view(d_input + node_range_begin[node_id],
+                                        node_range_end[node_id] - node_range_begin[node_id]);
+    return cudf::detail::default_hash<cudf::string_view>{}(field_name);
+  };
+  auto d_equal = [d_input          = d_input.data(),
+                  node_range_begin = d_tree.node_range_begin.data(),
+                  node_range_end   = d_tree.node_range_end.data()] __device__(auto node_id1,
+                                                                            auto node_id2) {
+    // TODO if node_category is used, then no need for transform later.
```

Comment:
Is this still relevant? It seems we don't run into the case given we're using `insert_if` and that `get_hash_value` is only getting invoked for field name nodes?

Reply:
not relevant. will remove this.